packages/core-data/src/types/README.md

```diff
@@ -0,0 +1,211 @@
+# Entity Records Types
+
+!THIS IS WORK IN PROGRESS! I used bits of the GitHub discussion as the first draft here
+
+## Overview
+
+The types in this directory are designed to support the following use-cases:
+
+* Context-aware type checks for entity records.
+* Static type checks for *edited* entity records, where certain fields become strings instead of objects.
+
+### Context-aware type checks for entity records
+
+WordPress REST API returns different responses based on the `context` query parameter, which typically is one of `view`, `edit`, or `embed`.
+
+For example, requesting `/wp/v2/posts/1?context=view` yields:
+
+```js
+{
+  "content": {
+    "protected": false,
+    "rendered": "\n<p>Welcome to WordPress. This is your first post. Edit or delete it, then start writing!</p>\n"
+  },
+  "title": {
+    "rendered": "Hello world!"
+  }
+  // other fields
+}
+```
+
+While requesting `/wp/v2/posts/1?context=edit`, yields:
+
+```js
+{
+  "content": {
+    "block_version": 1,
+    "protected": false,
+    "raw": "<!-- wp:paragraph -->\n<p>Welcome to WordPress. This is your first post. Edit or delete it, then start writing!</p>\n<!-- /wp:paragraph -->",
+    "rendered": "\n<p>Welcome to WordPress. This is your first post. Edit or delete it, then start writing!</p>\n"
+  },
+  "title": {
+    "raw": "Hello world!",
+    "rendered": "Hello world!"
+  }
+  // other fields
+}
+```
+
+And, finally, requesting `/wp/v2/posts/1?context=embed` yields:
+
+```js
+{
+  // Note content is missing
+  "title": {
+    "rendered": "Hello world!"
+  }
+  // other fields
+}
+```
+
+These contexts are supported by the core-data resolvers like `getEntityRecord()` and `getEntityRecords()` to retrieve the appropriate "flavor" of the data.
+
+The types describing different entity records must thus be aware of the relevant API context. This is implemented using the `Context` type parameter. For example, the implementation of the `Post` type resembles the following snippet:
+
+```ts
+import { OmitNevers } from "./helpers";
+
+interface BasePost<C extends Context> {
+	/**
+	 * A named status for the post.
+	 */
+	status: ContextualField< PostStatus, 'view' | 'edit', C >;
+
+	// ... other fields ...
+}
+
+export type Post<C extends Context> = OmitNevers<BasePost<C>>;
+```
+
+The `status` field of the `BasePost` is either a `PostStatus` when the Context is `view` or `edit`, or it's `never` when the Context is `embed`.
+
+The derived type `Post` is stripped of all the `never` fields, and so `Post<'embed'>` does not have the `status` type at all.
```

Comment:
we can probably get the same point across if all we showed were the `status` field line, omitting everything else.

we probably also don't need to go into details about the `never` values, instead describing the result of those internal type details: "The `status` field is a `PostStatus` when the requesting context is `view` or `edit`, but if requested with an `embed` context the field won't appear on the `Post` object at all."